File: src/EntityFx.CrystalMark/EntityFx.CrystalMark/ALU/Napierian.cpp
#include "Napierian.h"
#include <iostream>

int Napierian::bench_implementation(const std::atomic_bool& cancelled) {
    long long count = 0;
    unsigned short* a = new unsigned short[N + 1];
    unsigned short* t = new unsigned short[N + 1];

    for (; ; )
    {
        int m;
        unsigned int k;
        for (m = 0; m <= N; m++) {
            a[m] = t[m] = 0;
        }
        a[0] = 2;
        a[1] = t[1] = RADIX / 2;
        k = 3; m = 1;
        while ((m = divs(m, t, k, t)) <= N) {
            add(a, t, a);
            ++k;
        }

        count++;

        if (cancelled)
        {
            return (int)(count / 3.4 );
        }
    }
    return count;
}

int Napierian::divs(int m, unsigned short a[], unsigned int x, unsigned short b[])
{
    int i;
    unsigned long long t = 0;

    for (i = m; i <= N; i++)
    {
        t = (t << RADIXBITS) + a[i];
        b[i] = (unsigned short)(t / x);
        t %= x;
    }

    if (2 * t >= x)
    {
        for (i = N; (++b[i] & RADIX) == 1; i--)
        {
            b[i] &= RADIX - 1;
        }
    }
    return (b[m] != 0) ? m : (m + 1);
}

void Napierian::add(unsigned short a[], unsigned short b[], unsigned short c[])
{
    int i;
    unsigned short u = 0;

    for (i = N; i >= 0; i -= 2)
    {
        u += a[i] + b[i];
        c[i] = u & (RADIX - 1);
        u >>= RADIXBITS;

        u += a[i - 1] + b[i - 1];
        c[i - 1] = u & (RADIX - 1);
        u >>= RADIXBITS;
    }
}

string Napierian::get_name() {
    return "Napierian";
}

File: src/lua/alu/napierian.lua
Napierian = class(CrystalBenchmarkBase, function(fibonacci, writer)
    CrystalBenchmarkBase.init(fibonacci, writer, "Napierian")
end)

Napierian.N = 225

Napierian.RADIXBITS = 15;

Napierian.RADIX = 2 ^ Napierian.RADIXBITS;

function Napierian:benchImplementation()
    local count = 0
    local a = {}
    local t = {}

    while not self:isCancelled() do
        local m, k = 0, 0

        for ix = 1, Napierian.N, 1 do
            a[ix] = 0
            t[ix] = 0
        end

        a[1] = 2
        t[2] = math.floor(Napierian.RADIX / 2)
        a[2] = t[2]
        k = 3
        m = 1

        while true do
            m = Napierian:divs(m, t, k, t)
            if m > Napierian.N then
                break
            end
            Napierian:add(a, t, a)
            k = k + 1
        end

        count = count + 1
    end

    return count / 0.3197 * 4
end

function Napierian:divs(m, a, x, b)
    local i, t = 0, 0

    for i = m, Napierian.N, 1 do
        t = lshift(t, Napierian.RADIXBITS) + a[i]
        b[i] = math.floor(t / x)
        t = t % x
    end

    if 2 * t >= x then
        i = Napierian.N
        while true do
            b[i] = b[i] + 1
            if bitand(b[i], Napierian.RADIX) ~= 1 then
                break
            end
            b[i] = bitand(b[i], Napierian.RADIX - 1)
            i = i - 1
        end
    end

    if b[m] ~= 0 then
        return m
    else
        return m + 1
    end
end

function Napierian:add(a, b, c)
    local i, u = 0, 0
    for i = Napierian.N, 1, -1 do
        u = u + a[i] + b[i]
        c[i] = bitand(u, Napierian.RADIX - 1)
        u = rshift(u, Napierian.RADIXBITS)
    end
end
